File: 121.买卖股票的最佳时机.cpp
/*
 * @lc app=leetcode.cn id=121 lang=cpp
 *
 * [121] 买卖股票的最佳时机
 */

// @lc code=start

#include <vector>
#include <algorithm>

class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
        if(prices.empty()) return 0;
        int minB = prices[0];
        int maxL = 0;
        for(int i = 1; i < prices.size(); i++){
            minB = std::min(prices[i], minB);
            maxL = std::max(prices[i] - minB, maxL);
        }
        return maxL;
    }
};
// @lc code=end

File: 121.买卖股票的最佳时机.rs
/*
 * @lc app=leetcode.cn id=121 lang=rust
 *
 * [121] 买卖股票的最佳时机
 */

// @lc code=start
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut min_price = Option::<i32>::None;
        let mut result = 0;
        for price in prices {
            if let Some(min_price) = min_price.as_mut() {
                result = result.max(price - *min_price);
                *min_price = (*min_price).min(price);
            } else {
                min_price = Some(price);
            }
        }
        result
    }
}
// @lc code=end

